### PR TITLE
packagekit: Show Log of update actions

### PR DESCRIPTION
**Pull request by martinpitt**
In ApplyUpdates, remember all reported PackageKit actions instead of
just the last one, and display them in a log view in an expander below
the progress bar.

Automatically keep the update log scrolled to the bottom as long as it
is (almost) at the bottom already, so that without human interference it
always shows the latest information. Also scroll to the bottom when the
Expander gets opened; add a new `onExpand()` callback for this purpose.
    
Check this in TestUpdates.testBasic(). For that the test needs to
enforce an ordering on package updates, so add dependencies to the
generated test packages.

### CODE REVIEW

**Comment by martinpitt** [comment]
This is @garrett's design:
design

This is a real-life (Fedora 26) sceenshot from my system with this PR:

updates-log

This keeps the log auto-scrolled to the bottom if it already is (almost) at the bottom, when adding a new entry.

@garrett, can you please do a design review of this? Thanks!

**Comment by martinpitt** [comment]
Compared to the design, the update log is wider. I kept the same width (and components) as for the update history, so that the two have the same layout.

File: pkg/packagekit/updates.jsx [hunk 1]
```diff
@@ -319,40 +319,64 @@ class ApplyUpdates extends React.Component {
...
-                        this.setState({ curPackage: pfields[0] + " " + pfields[1],
-                                        curStatus: info,
+                        let newActions = this.state.actions.slice();
+                        let statusStr = PK_STATUS_STRINGS[info || PK_STATUS_ENUM_UPDATE] ||
+                            PK_STATUS_STRINGS[PK_STATUS_ENUM_UPDATE];
+                        newActions.push([statusStr, pfields[0] + " " + pfields[1]]);
+
+                        let log = document.getElementById("update-log");
+                        let atBottom = false;
+                        if (log)
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: `if` block is longer than one line, so please add braces

File: pkg/packagekit/updates.jsx [hunk 2]
```diff
@@ -319,40 +319,64 @@ class ApplyUpdates extends React.Component {
...
-            action = (
-                <span>
-                    <strong>{ PK_STATUS_STRINGS[this.state.curStatus || PK_STATUS_ENUM_UPDATE] || PK_STATUS_STRINGS[PK_STATUS_ENUM_UPDATE] }</strong>
-                    &nbsp;{this.state.curPackage}
-                </span>
-            );
-        else
-            action = _("Initializing...");
+        if (this.state.actions.length > 0) {
+            let lastAction = this.state.actions[this.state.actions.length-1];
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: binary operators [should be separated from operands by spaces](https://github.com/cockpit-project/cockpit/wiki/Cockpit-Coding-Guidelines#jsstyle)
`let lastAction = this.state.actions[this.state.actions.length - 1];`

File: test/verify/check-packagekit [hunk 3]
```diff
@@ -117,7 +117,17 @@ class TestUpdates(MachineCase):
         # Cancel button should eventually get disabled
         b.wait_present(".content-header-extra td button:disabled")
 
-        # should have succeeded and show restart page; cancel
+        # update log is not visible by default
```
Comment (dperpeet) [inline test/verify/check-packagekit]: not visible and not present can have different meaning in these tests - perhaps the comment should be clearer?

File: pkg/packagekit/updates.jsx (hunk 1 above)
Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: @dperpeet: Fixed your three issues, thanks for the review!

**Comment by martinpitt** [comment]
Tagging priority due to the current truckload of naughty cleanup PRs, to get feedback in some finite time.

**Comment by martinpitt** [comment]
[Software Updates: upgrade log, layout, RHEL registration](https://trello.com/c/w0weB970/534-software-updates-upgrade-log-layout-rhel-registration)

**Comment by garrett** [comment]
OK! So I have this set up locally and downgraded and have been playing with it a bit. I have noticed a few issues:

- Unless the scrollable area is already at the bottom, it doesn't scroll to the latest updated package (it should scroll to the latest at every update)
- Action word tense doesn't change — "Updating" should change to "Updated", "Setting up" should become "Set up", etc.

I'll play around with it a bit more — aside from the above two issues, everything else looks good to me so far.

**Comment by garrett** [comment]
I realized how I triggered the issue: If you expand, collapse, and expand again, the scrollable area is at the top and doesn't scroll down. The fix is to always scroll to the bottom on every expand.

This wouldn't interfere with someone viewing the logs, but would work when someone plays with the element (as people often do, especially when they're waiting for something to happen).

**Comment by martinpitt** [comment]
@garrett: Fixed, it now always scrolls to the bottom when opening the expander.

**Comment by dperpeet** [comment]
If the window isn't high enough, I need to scroll down to read the end even though there is empty space at the top of the page. This isn't a big bother though and certainly not a regression.

File: test/verify/check-packagekit [hunk 4]
```diff
@@ -67,7 +67,7 @@ class TestUpdates(MachineCase):
 
         # create two updates
         self.createPackage("vanilla", "1.0", "1", install=True)
-        self.createPackage("vanilla", "1.0", "2")
+        self.createPackage("vanilla", "1.0", "2", postinst='sleep 10')
```
Comment (dperpeet) [inline test/verify/check-packagekit]: we don't want to introduce a fixed sleep time here again

Comment (martinpitt) [inline test/verify/check-packagekit]: I'm glad you introduced the finish-pk lockfile ealier; moving to this uncovered a race condition in the test, as we can't assume that chocolate gets installed before vanilla. But the test relies on this, so I added dependencies to the test packages. Now both the ordering and the sleep races are fixed.

Force-pushed the update.

File: pkg/packagekit/updates.jsx [hunk 5]
```diff
@@ -301,7 +314,7 @@ function UpdateHistory(props) {
 class ApplyUpdates extends React.Component {
     constructor() {
         super();
-        this.state = { percentage: 0, timeRemaining: null, curStatus: null, curPackage: null };
+        this.state = { percentage: 0, timeRemaining: null, actions: [] };
```
Comment (dperpeet) [inline pkg/packagekit/updates.jsx]: Can you add a comment on how `actions` is filled?
Do you think it makes sense to use an object for each entry instead? `action[0]` and `action[1]` are a lot more difficult to read and debug than `action['packageInfo']`

Comment (martinpitt) [inline pkg/packagekit/updates.jsx]: Indeed. I also broke the long lines a bit, see the [interdiff](https://paste.fedoraproject.org/paste/ss0h9Sn8okMExWpmp3kSdQ) to the previous version of this PR. Pushed.

**Pull request merged**